gpu_sysman plugin update to v6.0 API + new labels + "raw" metrics + output options

eero-t opened the pull request.

eero-t committed:
gpu_sysman: remove "GPU-" prefix from name and add it "pci_pdf" label

Also rename GPU struct "name" member to more explicit "pci_bdf".

This allowed simplifying the code slightly.

Sysman API supports nowadays also other devices than GPUs, so prefix
is removed to to simplify code and to be more future-proof:
https://spec.oneapi.io/level-zero/latest/core/api.html#_CPPv416ze_device_type_t

(Plugin will still query only GPU devices from Sysman though.)

eero-t committed:
gpu_sysman: use sizeof(*var) rather than sizeof(vartype) in var=calloc(...)

Except for gpu_subarray_alloc(), all allocs are done with calloc().
This way correctness of all of them is easy to check just by grepping
for calloc (especially now that clang-format does not wrap those lines
any more), and reviewing gpu_subarray_alloc().

eero-t committed:
gpu_sysman: split metric properties from their names to separate labels

Following labels are used:
- sub_dev: subdevice ID (unsigned integer)
- location: e.g. "gpu" / "memory"
- type: e.g. "request" / "actual"
- direction: "read" / "write"

Additionally:

* Two location label values were fixed

* GPU engine indeces are now per engine type
  (instead of single index being used for all types)

* All metric family and label names have been changed to use
  underscores instead of dashes to separate words, as required by
  Prometheus i.e. collectd does not need to convert them any more:
  https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels

eero-t committed:
gpu_sysman: update test code to handle metrics split with labels

NOTE: providing NULL as label value to delete it is NOT supported.
Test code will assert on labels with NULL values.

eero-t committed:
gpu_sysman: fix test code for "pci_bdf" added to metrics family

- do not add "pci_pdf" to metric name for matching
- fix for adding metric labels to family copies of them

eero-t committed:
gpu_sysman: improvements to reported metrics

* Fix memory "type" label overwrite

* Replace "free" memory metric with "memory_usage_ratio" one,
  and rename "memory_bytes" to "memory_used_bytes" metric

* Split metric value aggregate function name to a separate
  "function" label

* Have metric family declares always in same place in code

* Avoid both setting metric labels, and reporting empty metrics,
  when higher internal sampling rate is used or there are L0
  errors

eero-t committed:
gpu_sysman: update tests for sysman plugin changes

* Add "memory_usage_ratio" checks

* Update validation for metrics that can be sampled at higher
  rate i.e. have now the new aggregate function label

* With empty metrics avoided, dispatch mock-up can assert on them

* With extra L0 calls being skipped when not needed, number of calls
  can differ between query rounds:
  - refactor multi-sampling test to handle count changes
  - change error handing checks to be done in single-sampled mode

* Debug output is needed to debug triggered multisample asserts,
  so do that when assert would have been triggered, then abort

eero-t committed:
gpu_sysman: add help information for all metric families

And document why const-qual cast is safe, and why GCC does
not warn about other assignments to .name & .help members.

eero-t committed:
gpu_sysman: option to disable utilization metrics for single engines

More powerful GPUs can have a large number of engines of given type,
but user may be interested only on the higher level engine groups
utilization.

"DisableEngineSingle" option allows skipping individual engine metrics.

eero-t committed:
gpu_sysman: option for specifying metrics output type

This can be used to speciify whether output metrics values will be
raw, derived or both.

This commit add support just for the configuration option itself,
adding / changing metrics to use it happens in next commit.

eero-t committed:
gpu_sysman: optional raw metrics output for already supported metrics

This adds new counter type metrics for:
* memory bandwidth
* frequency throttle time
* engine execution time (activity)
* energy usage

Because collecd internally handles counters as integers, all units
cannot be ones recommended by Prometheus, but microseconds and
microjoules reported by Sysman.

eero-t committed:
gpu_sysman: minimal v6 API support + add units to metric names

Prometheus & OpenMetrics require metric names to be suffixed by the
metric unit, and ratios (0-1) to be used instead of percentages
(0-100).

eero-t committed:
gpu_sysman: update test code for minimal v6 API support + new metric names

There's now also support for multiple metrics per family although they
are not used yet. "sstrncpy" is not needed any more.

byako reviewed (approved):
LGTM

uniemimu commented on src/gpu_sysman.c:
```diff
@@ -591,19 +613,37 @@ static int gpu_init(void) {
+                       const char *type, const char *subdev, double value) {
+  metric_family_t fam = {
+      .type = METRIC_TYPE_COUNTER,
+      .name = (char *)name,
```
I hope these don't get written to, as in case if the incoming is truly const, writes result in undefined behavior

eero-t replied:
Compilers puts all literal strings to read-only section by default, so in actuality they are const strings.  Assigning them to const char*, and propagating them e.g. to ras_submit() as such, is the correct thing to do, regardless of the enabled compiler warnings [1].

The annoying thing here is that metric_family_t.name in collectd v6 is non-const string.  This is because collectd offers also functions for allocation and freeing metric families, and for this purpose the .name member cannot be const.

My code does not use / need the metric family alloc / free functions, only the metrics processing ones, so explicitly casting the literal name + help strings to non-const is safe.  Besides, the same cast happens implicitly [1] in the many places where literal strings are assigned to metric_family_t.name directly.

[1] GCC does not do const checks for literal strings even with "-Wall -Wextra", one needs to explicitly use "-Write-strings" before GCC emits warnings about their assignments.  "-Write-strings" GCC option would cause a large number of warnings about "metric_family_t.name" usage elsewhere than in this function, *and* in all the other collectd plugins though.

Allocating all the name & help strings just to be able to enable "-Wwrite-strings" warning is not worth the overhead though.

uniemimu commented on src/gpu_sysman.c:
```diff
@@ -824,24 +903,21 @@ static bool gpu_mems(gpu_device_t *gpu, unsigned int cache_idx) {
+      mem_used = mem_size - mem_free;
+      metric.value.gauge = mem_used;
+      metric_family_metric_append(&fam_bytes, metric);
+      metric.value.gauge = mem_used / mem_size;
```
is the mem_size guaranteed to be > 0 always?

eero-t replied:
According to spec it seems to be:
* https://spec.oneapi.io/level-zero/latest/sysman/api.html#zes-mem-state-t
* https://spec.oneapi.io/level-zero/latest/sysman/api.html#zes-mem-properties-t

> 
> zes_mem_state_t: uint64_t size:  [out] The total allocatable memory in bytes (can be less than zes_mem_properties_t.physicalSize)
> ...
> zes_mem_properties_t: uint64_t physicalSize: [out] Physical memory size in bytes. A value of 0 indicates that this property is not known. However, a call to zesMemoryGetState() will correctly return the total size of usable memory.

But I guess there could be either kernel or sysman bugs...

eero-t replied:
Added check for this.

uniemimu commented on src/gpu_sysman.c:
```diff
@@ -895,58 +983,87 @@ static bool gpu_mems_bw(gpu_device_t *gpu) {
       uint64_t timediff = bw.timestamp - old->timestamp;
-      double factor = 100.0 * 1.0e6 / (bw.maxBandwidth * timediff);
-      char vname[64];
+      double factor = 1.0e6 / (bw.maxBandwidth * timediff);
```
can we rest assured that there is no possibility for a div-by-zero here?

eero-t replied:
The added commit handles that by comparing timestamps, and using old->maxBandwidth.

uniemimu commented on src/gpu_sysman.c:
```diff
@@ -1112,40 +1250,68 @@ static bool gpu_freqs_throttle(gpu_device_t *gpu) {
-      gpu_submit(gpu->name, "percent", vname, 100.0 * throttled);
+    if (old->timestamp && (config.output & OUTPUT_DERIVED)) {
+      /* micro seconds => throttle ratio */
+      metric.value.gauge = (throttle.throttleTime - old->throttleTime) /
```
no div-check here either. It probably works fine if the timestamps are guaranteed to be good enough, but I don't know that

eero-t replied:
I handled that by skipping also timestamp wrap-around, as both zero time interval and wrapped one would be bogus.

uniemimu reviewed (commented):
apart from the several division operations, and them const removals with casts, I think this looks fine. If you ensure that the divs can't be zero, and that them removed consts don't get written to, I'll say LGTM.

eero-t committed:
gpu_sysman: skip metrics with div-by-zero or time wrap around issues

Zero time intervals or max bandwidth would cause div-by-zero issues
and (very rare) time wrap around would cause bogus metric value.
Skip all of them.

uniemimu commented:
LGTM

eero-t committed:
gpu_sysman: fix test code -Wpedantic + -Wcast-qual warnings

eero-t commented:
Rebased the test code debug output change commit to one adding it initially, rebased clang-format white-space changes to appropriate commits + added comment commit explaining the const casting.

eero-t committed:
gpu_sysman: add 'sub_dev' and 'type' labels only when needed

Empty label equals to a missing one, and Prometheus queries can check
for non-existence of a label, so let's just skip empty / unneeded ones.

Main difference to earlier is that LevelZero error categories that
provide non-zero values only for uncorrectable type (according to
spec), are now without a type label. Correctable i.e. zero metrics for
those categories were skipped already earlier.

eero-t committed:
Add "dev_file" label support

And contrib/format.sh include re-order.

"dev_file" support is behind a define (enabled by default) because it
needs functions that are only part of POSIX, not C99.

Intel Kubernetes GPU plugin uses primary GPU node device file names
(card0, card1...) as its GPU identifiers.  This new label helps in
mapping Kubernetes custom metrics to them.

eero-t commented:
@uniemimu and @byako, could you review the 2 new commits, especially the last "dev_file" label one (intended to help with k8s custom rules)?

uniemimu reviewed (commented):
lgtm

byako reviewed (commented):
LGTM

eero-t commented:
To document my chat with Alexey: I'm leaving handling of when ADD_DEV_FILE define is enabled until upstream comments on whether they: 1) want that Kubernetes Intel GPU plugin specific feature to be disabled in upstream, 2) a configure option for enabling it, or 3) configure checks for the required POSIX functions before "dev_file" label feature gets enabled.

eero-t commented:
The reason for last two force pushes is that I'm using the reviewed branch for upstreaming (https://github.com/collectd/collectd/pull/3968), and I rebased that to:
* Move separate Prometheus PR (needed for testing the new gpu_sysman plugin in k8s cluster) out from a branch beneath it, to a new new dev branch above it
* Apply contrib/format.sh changes that collectd CI told to run.  I've been using local (LLVM 13) clang-format, but changes it does differ slightly from what remote clang-format (called by format.sh) does
 
Unfortunately contrib/format.sh changes did not satisfy whichever clang-format version / settings collectd CI itself runs => expect further pushes once upstream tells what actually should be changed in regards to code formatting.

eero-t commented:
This was rebased on latest upstream collectd-6.0 branch, so that "gpu_sysman-6.0" branch based on this branch, gets CI fixes from upstream => upstream's own bugs do not fail checks for upstream "gpu_sysman-6.0" branch pull request.

eero-t committed:
Move test defines from Sysman plugin to its test code

And document with what GCC warning options the code is tested / passes.

eero-t committed:
Change strcpy() in Sysman plugin to sstrncpy()

While for plugin that change does not really help (as target buffer is
always larger than source), for test code it is useful. And it shuts
up less capabable static checking tools than GCC.

As test code cannot use existing collectd functionality for this (test
code needs modified versions of some collectd functions, and all
collectd code does not pass GCC warnings I use), sstrncpy() is copied
to test code.

For test code there's also a fix to size given for snprintf(), and
removal of redundant string termination for modified plugin_log() copy
(vsnprintf() already terminates string).

eero-t committed:
Pass clang-format check for gpu_sysman_test.c comments

eero-t committed:
Add scalloc() wrapper similar to smalloc() to common utils

scalloc() wraps calloc() with exit on alloc failure,
similarly to what smalloc() does for malloc().

eero-t committed:
Replace Sysman plugin alloc+assert calls with smalloc/scalloc

If asserts were disabled, allocation failures would result in collectd
memory errors => replace alloc+assert in the plugin with collectd
smalloc/scalloc wrappers that exits after logging allocation error.

Downsides are that this does not invoke debugger (which could be in a
different control group with plenty of memory), nor tell where / what
allocation failed, like enabled assert would, so test code variants of
the wrappers still do asserts.

eero-t committed:
Pass clang-format check for gpu_sysman_test.c

eero-t commented:
Closing this "gpu_sysman-6.0" branch review PR, upstream finally reviewed & merged sysman plugin.
